### PR TITLE
Split public and private question data.

### PR DESCRIPTION
**Pull request by rockon999**
### Summary
- Private data is stored in a subcollection/doc at `/questions/{question}/content/private`
- Selectively load question data when isTA is flagged

(more coming soon)

What currently works:
- Student view

What is a WIP:
- TA view

### CODE REVIEW

File: src/components/includes/SessionQuestionsContainer.tsx [hunk 1]
```diff
@@ -38,7 +40,18 @@ const SessionQuestionsContainer = (props: Props) => {
...
 
-    const allQuestions = props.questions;
+    const initQuestions = props.questions;
+    const [allQuestions, setAllQuestions] = React.useState<FireEntry[]>([]);
+
+    React.useEffect(() => {
+        if (props.isTA) {
+            Promise.all(allQuestions.map(q => firestore.collection('questions').doc(q.questionId).get())).then(q => {
+                setAllQuestions(q.map(q => ({ questionId: q.id, ...q.data() } as unknown as FireQuestion)));
+            });
```
Comment (SamChou19815) [inline src/components/includes/SessionQuestionsContainer.tsx:50]: This will be a terrible `O(n)` number of IOs. We should fetch from questions collection directly instead.

File: src/components/types/fireData.d.ts [hunk 2]
```diff
@@ -70,6 +70,16 @@ interface FireUser {
...
+interface FireSlot {
+    askerId: string;
+    sessionId: string;
+    status: 'assigned' | 'resolved' | 'retracted' | 'unresolved' | 'no-show';
+    timeAddressed?: FireTimestamp;
+    timeEntered: FireTimestamp;
+    questionId: string;
+    questionRef: string;
+}
+
```
Comment (SamChou19815) [inline src/components/types/fireData.d.ts:82]: We should rename it as `FireQuestionSlot` and the collection as `questionSlots` for clarity.

File: src/components/includes/SessionQuestion.tsx [hunk 3]
```diff
@@ -119,42 +134,62 @@ class SessionQuestion extends React.Component<Props, State> {
...
+        firestore.doc(`questions/${this.props.entry.questionId}`).update({
             status: 'resolved',
             timeAddressed: firebase.firestore.Timestamp.now()
         });
+
+        const slot = firestore.collection('slots').doc(this.props.entry.questionId);
+
+        slot.update({
+            status: 'resolved'
+        });
```
Comment (SamChou19815) [inline src/components/includes/SessionQuestion.tsx:171]: Use `firestore.batch` to batch the updates

File: src/components/includes/CalendarView.tsx [hunk 4]
```diff
@@ -27,10 +27,21 @@ export default ({ session, sessionCallback, course, user }: Props) => {
...
+        const end = session.endTime.toDate();
+        // TODO This doesn't handle timezones quite right but 
+        // works better than the other somehow...
+        return (selectedDate.getUTCFullYear() === start.getUTCFullYear() &&
+            selectedDate.getUTCMonth() === start.getUTCMonth() &&
+            selectedDate.getUTCDate() === start.getUTCDate()) ||
+            (selectedDate.getUTCFullYear() === end.getUTCFullYear() &&
+                selectedDate.getUTCMonth() === end.getUTCMonth() &&
+                selectedDate.getUTCDate() === end.getUTCDate());
+    });
```
Comment (SamChou19815) [inline src/components/includes/CalendarView.tsx:44]: What's the rationale for the change?

File: src/components/includes/AddQuestion.tsx [hunk 5]
```diff
@@ -132,7 +132,26 @@ class AddQuestion extends React.Component<Props, State> {
...
+                            status,
+                            questionId: question.id,
+                            timeEntered,
+                            questionRef: question.id
+                        };
+
+                        firestore.collection('slots').doc(question.id).set(newSlot).catch(err => console.log(err));
+                    }).catch(err => console.log(err)) ;
+                });
             this.setState({ redirect: true });
```
Comment (SamChou19815) [inline src/components/includes/AddQuestion.tsx:155]: We should pre-generate an ID using `firestore.collection('questions').doc().id`, and use the same id for both.
Then this can be done in a single batch operation.

File: src/components/includes/SessionQuestionsContainer.tsx (hunk 1 above)
Comment (rockon999) [reply src/components/includes/SessionQuestionsContainer.tsx]: Agreed, luckily this is gone. But we have a similar issue in fetching all the private data. In that case, I'm thinking a query is possible.

File: src/components/includes/CalendarView.tsx (hunk 4 above)
Comment (rockon999) [reply src/components/includes/CalendarView.tsx]: See the updated code, it is actually correct. The current filtering logic is completely broken for me in the late night time hours.

File: src/components/includes/SessionQuestionsContainer.tsx (hunk 1 above)
Comment (SamChou19815) [reply src/components/includes/SessionQuestionsContainer.tsx]: @rockon999 I don't think subcollection query can work. The only possible query that can be done on subcollection is collection group query: https://firebase.google.com/docs/firestore/query-data/queries#collection-group-query
, but it doesn't solve our problem.
